tools/release_manifest.nim

```diff
@@ -0,0 +1,564 @@
+  ##
+  ## :triplet:
+  ##   The triplet it is targeting
+  if file.len == 0:
+    raise newException(ValueError):
+      "The file name must not be empty"
+
+  if version.len == 0:
+    raise newException(ValueError):
+      "The version must not be empty"
+
+  if file in d.file:
+    raise newException(ValueError):
+      "Artifact already exist in the database: " & file
+
+  if sha256.len != Sha256StrLength:
+    raise newException(ValueError):
+      "Invalid length of (" & $sha256.len & ") for sha256 checksum: " & sha256
+
+  if triplet in d.triplet:
+    let id = d.triplet.find(triplet)
+    raise newException(ValueError):
+      if triplet != SourceTriplet:
+        "Could not add artifact (" & file & "): artifact for " & triplet &
+        " already exist: " & d.file[id]
+      else:
+        "Could not add artifact (" & file & "):" &
+        " a source archive has already been added: " & d.file[id]
+
+  # If no version is in the database (empty database)
+  if d.version.len == 0:
+    # Set the version to the artifact's version
+    d.version = version
+  # If there is a version and it doesn't match the artifact's
+  elif d.version != version:
+    raise newException(ValueError):
+      "The artifact to be added is of a different version " & version &
+      " from the database version " & d.version
+
+  d.file.add file
+  d.sha256.add sha256
+  d.triplet.add triplet
+
+func serialize(d: Database): JsonNode =
+  ## Convert the database into a manifest to be used by tools.
+  ##
+  ## For the schema of the resulting JSON document, refer to the
+  ## `ReleaseManifest` structure.
+  var manifest = ReleaseManifest(
+    manifestVersion: ManifestVersion,
+    version: d.version
+  )
+
+  for i in 0 ..< d.file.len:
+    if d.triplet[i] == SourceTriplet:
+      manifest.source = ArtifactDataV0(
+        name: d.file[i],
+        sha256: d.sha256[i]
+      )
+    else:
+      manifest.binaries.add:
+        BinaryArtifactDataV0(
+          target: d.triplet[i],
+          name: d.file[i],
+          sha256: d.sha256[i]
+        )
+
+  result = %manifest
+
+func deserialize(j: JsonNode): Database =
+  ## Convert the manifest for tools into a database.
+  if j["manifestVersion"].getInt(-1) != 0:
+    raise newException(ValueError):
+      "Unknown manifest version: " & $j["manifestVersion"].getInt(-1)
+
+  # Currently only manifest version 0 is supported.
+  let manifest = j.to(ReleaseManifestV0)
+
+  result.version = manifest.version
+  # Add the source artifact if available
+  if manifest.source.name.len > 0:
+    result.addArtifact(
+      manifest.source.name, manifest.source.sha256, manifest.version,
+      SourceTriplet
+    )
+
+  # Add all artifacts
+  for artifact in manifest.binaries.items:
+    result.addArtifact(
+      artifact.name, artifact.sha256, manifest.version, artifact.target
+    )
+
+func toTriplet(os, cpu: string): string =
+  ## Convert Nim's os/cpu pair into a LLVM-style target triplet.
```

Comment:
Did you have any go to reference links you used for triplet study when devising this?

If so please link then in comments for those modifying this in the future.